File: src/core/hashing.ts
/**
 * Hashing module for JnxBetaSec.
 */

import * as fs from "fs"
import * as crypto from "crypto"
import { promisify } from "util"
import type { HashAlgorithm } from "../types"

const fsReadFilePromise = promisify(fs.readFile)
const fsStatPromise = promisify(fs.stat)

export class Hashing {
  private readonly SUPPORTED_ALGORITHMS: Record<string, () => crypto.Hash> = {
    md5: () => crypto.createHash("md5"),
    sha1: () => crypto.createHash("sha1"),
    sha256: () => crypto.createHash("sha256"),
    sha384: () => crypto.createHash("sha384"),
    sha512: () => crypto.createHash("sha512"),
    blake2b512: () => crypto.createHash("blake2b512"),
    blake2s256: () => crypto.createHash("blake2s256"),
  };
  
  constructor() {
    // No initialization needed
  }

  /**
   * Generate a hash for a file.
   *
   * @param filePath Path to the file
   * @param algorithm Hash algorithm to use
   * @param chunkSize Size of chunks to read from file
   * @returns Hexadecimal hash string
   */
  public async hashFile(filePath: string, algorithm: HashAlgorithm = "sha256", chunkSize = 8192): Promise<string> {
    try {
      await fsStatPromise(filePath)
    } catch (e) {
      throw new Error(`File not found: ${filePath}`)
    }

    if (!this.SUPPORTED_ALGORITHMS[algorithm.toLowerCase()]) {
      throw new Error(
        `Unsupported algorithm: ${algorithm}. Supported algorithms: ${Object.keys(this.SUPPORTED_ALGORITHMS).join(", ")}`,
      )
    }

    return new Promise<string>((resolve, reject) => {
      const hash = this.SUPPORTED_ALGORITHMS[algorithm.toLowerCase()]()
      const stream = fs.createReadStream(filePath, { highWaterMark: chunkSize })

      stream.on("data", (chunk) => {
        hash.update(chunk)
      })

      stream.on("end", () => {
        resolve(hash.digest("hex"))
      })

      stream.on("error", (err) => {
        reject(err)
      })
    })
  }

  /**
   * Verify a file against an expected hash.
   *
   * @param filePath Path to the file
   * @param expectedHash Expected hash value
   * @param algorithm Hash algorithm to use
   * @returns True if the hash matches, False otherwise
   */
  public async verifyFile(
    filePath: string,
    expectedHash: string,
    algorithm: HashAlgorithm = "sha256",
  ): Promise<boolean> {
    const actualHash = await this.hashFile(filePath, algorithm)
    return actualHash.toLowerCase() === expectedHash.toLowerCase()
  }

  /**
   * Generate a hash for a string.
   *
   * @param inputString String to hash
   * @param algorithm Hash algorithm to use
   * @returns Hexadecimal hash string
   */
  public hashString(inputString: string, algorithm: HashAlgorithm = "sha256"): string {
    if (!this.SUPPORTED_ALGORITHMS[algorithm.toLowerCase()]) {
      throw new Error(
        `Unsupported algorithm: ${algorithm}. Supported algorithms: ${Object.keys(this.SUPPORTED_ALGORITHMS).join(", ")}`,
      )
    }

    const hash = this.SUPPORTED_ALGORITHMS[algorithm.toLowerCase()]()
    hash.update(inputString)
    return hash.digest("hex")
  }
}

File: jnxbetasec/core/hashing.py
"""
Hashing module for JnxBetaSec.
"""

import os
import hashlib
import logging
from typing import Dict, Optional
from pathlib import Path

logger = logging.getLogger("jnxbetasec.hashing")

class Hashing:
    """
    JnxBetaSec Hashing System
    
    Provides secure file hashing and verification capabilities.
    """
    
    SUPPORTED_ALGORITHMS = {
        "md5": hashlib.md5,
        "sha1": hashlib.sha1,
        "sha256": hashlib.sha256,
        "sha384": hashlib.sha384,
        "sha512": hashlib.sha512,
        "blake2b": hashlib.blake2b,
        "blake2s": hashlib.blake2s,
    }
    
    def __init__(self):
        """Initialize the JnxBetaSec hashing system."""
        pass
    
    def hash_file(self, file_path: str, algorithm: str = "sha256", chunk_size: int = 8192) -> str:
        """
        Generate a hash for a file.
        
        Args:
            file_path: Path to the file
            algorithm: Hash algorithm to use
            chunk_size: Size of chunks to read from file
            
        Returns:
            Hexadecimal hash string
        """
        file_path = Path(file_path)
        if not file_path.exists():
            raise FileNotFoundError(f"File not found: {file_path}")
        
        if algorithm.lower() not in self.SUPPORTED_ALGORITHMS:
            raise ValueError(f"Unsupported algorithm: {algorithm}. Supported algorithms: {', '.join(self.SUPPORTED_ALGORITHMS.keys())}")
        
        hash_func = self.SUPPORTED_ALGORITHMS[algorithm.lower()]()
        
        with open(file_path, "rb") as f:
            for chunk in iter(lambda: f.read(chunk_size), b""):
                hash_func.update(chunk)
        
        return hash_func.hexdigest()
    
    def verify_file(self, file_path: str, expected_hash: str, algorithm: str = "sha256") -> bool:
        """
        Verify a file against an expected hash.
        
        Args:
            file_path: Path to the file
            expected_hash: Expected hash value
            algorithm: Hash algorithm to use
            
        Returns:
            True if the hash matches, False otherwise
        """
        actual_hash = self.hash_file(file_path, algorithm)
        return actual_hash.lower() == expected_hash.lower()
    
    def hash_string(self, input_string: str, algorithm: str = "sha256") -> str:
        """
        Generate a hash for a string.
        
        Args:
            input_string: String to hash
            algorithm: Hash algorithm to use
            
        Returns:
            Hexadecimal hash string
        """
        if algorithm.lower() not in self.SUPPORTED_ALGORITHMS:
            raise ValueError(f"Unsupported algorithm: {algorithm}. Supported algorithms: {', '.join(self.SUPPORTED_ALGORITHMS.keys())}")
        
        hash_func = self.SUPPORTED_ALGORITHMS[algorithm.lower()]()
        hash_func.update(input_string.encode('utf-8'))
        
        return hash_func.hexdigest()

